src/Extensions/IndexTree.cs

```diff
@@ -0,0 +1,132 @@
+﻿using System;
+using System.Linq;
+
+namespace Wangkanai.Detection.Extensions
+{
+    public readonly struct IndexTree
```

Comment:
Very interest utilities class you have got there.

Reply:
This structure will combine all keywords into simple tree to walk it while searching, I have plans to make this even more efficient by integrating KMP (Knuth–Morris–Pratt) search algorithm.